proof/refine/ARM/Tcb_R.thy

```diff
@@ -2723,20 +2723,120 @@ lemma decodeSetSchedParams_wf[wp]:
      list_all2 (\<lambda>(c, sl) (c', sl'). cap_relation c c' \<and> sl' = cte_map sl) extras extras' \<rbrakk> \<Longrightarrow>
    corres (ser \<oplus> tcbinv_relation)
-       (cur_tcb and valid_etcbs and (\<lambda>s. \<forall>x \<in> set extras. s \<turnstile> (fst x)))
+       (cur_tcb and (\<lambda>s. \<forall>x \<in> set extras. s \<turnstile> (fst x)))
        (invs' and (\<lambda>s. \<forall>x \<in> set extras'. s \<turnstile>' (fst x)))
        (decode_set_sched_params args cap slot extras)
        (decodeSetSchedParams args cap' slot' extras')"
   apply (simp add: decode_set_sched_params_def decodeSetSchedParams_def)
   apply (cases "length args < 2")
    apply (clarsimp split: list.split)
-  apply (cases "length extras < 1")
+  apply (cases "length extras < 3")
    apply (clarsimp split: list.split simp: list_all2_Cons2)
   apply (clarsimp simp: list_all2_Cons1 neq_Nil_conv val_le_length_Cons linorder_not_less)
+  apply (rule corres_split_eqrE)
+     apply (rule corres_split_norE[OF _ check_prio_corres])
+       apply (rule corres_split_norE[OF _ check_prio_corres])
+         apply (case_tac cap; clarsimp)
+         apply (rename_tac a b c d e sc_cap f g fh i j k l sc_cap' fh' n auth_tcb tcb_ptr)
+         apply (prop_tac "corres (ser \<oplus> (\<lambda>r r'. tc_new_sc r = Some r')) P P'
+                                 (decode_update_sc (cap.ThreadCap tcb_ptr) slot sc_cap)
+                                 (case sc_cap' of capability.NullCap \<Rightarrow>
+                                      doE curTcbPtr <- withoutFailure getCurThread;
+                                          whenE (tcb_ptr = curTcbPtr) $
+                                            throw Fault_H.syscall_error.IllegalOperation;
+                                          returnOk Nothing
+                                      odE
+                                  | capability.SchedContextCap scPtr x \<Rightarrow> doE tcbSc <- withoutFailure $ threadGet tcbSchedContext tcb_ptr;
+                                                             whenE (tcbSc \<noteq> Nothing) $ throw Fault_H.syscall_error.IllegalOperation;
+                                                             scTcb <- withoutFailure $ mapScPtr scPtr scTCB;
+                                                             whenE (scTcb \<noteq> Nothing) $ throw Fault_H.syscall_error.IllegalOperation;
+                                                             blockedAndUnreleased <- withoutFailure $ isBlocked tcb_ptr `~andM~` (liftM Not $ scReleased scPtr);
+                                                             whenE blockedAndUnreleased $ throw Fault_H.syscall_error.IllegalOperation;
+                                                             returnOk $ Just scPtr
+                                                         odE
+                                  | _ \<Rightarrow> throw $ Fault_H.syscall_error.InvalidCapability 2)")
```

Comment:
This is a big `prop_tac`. It probably makes sense to have this as a separate lemma? Or any reason you choose not to?